sdk/cosmos/azure-cosmos/azure/cosmos/_cosmos_client_connection.py

```diff
@@ -1765,6 +1769,83 @@ def PatchItem(self, document_link, operations, options=None, **kwargs):
         self._UpdateSessionIfRequired(headers, result, self.last_response_headers)
         return result
 
+    def Bulk(self, collection_link, operations, options=None, **kwargs):
+        """Executes the given operations in bulk.
+
+        :param str collection_link: The link to the collection
+        :param list operations: The operations for the bulk request.
+        :param dict options: The request options for the request.
+
+        :return:
+            The result of the bulk operation.
+        :rtype:
+            dict
+
+        """
+        if options is None:
+            options = {}
+
+        pk_ranges = self._ReadPartitionKeyRanges(collection_link)
+        batches = []
+        for pk_range in pk_ranges:
+            batch = {
+                "min": pk_range.get("minInclusive"),
+                "max": pk_range.get("maxExclusive"),
+                "range_id": pk_range.get("id"),
+                "indexes": [],
+                "operations": []
+            }
+            batches.append(batch)
+        for operation in operations:
+            partition_key_value = operation.get("partitionKey", None)
+            if partition_key_value is None:
+                operation = self._AddPartitionKey(collection_link, operation.get("resourceBody"), operation)
+                partition_key_value = operation.get("partitionKey", None)
+                if partition_key_value is None:
+                    raise AttributeError("Bulk operation must have a partition key specified.")
+            operation["partitionKey"] = json.dumps([partition_key_value])
+            pk_definition = self._get_partition_key_definition(collection_link)
+            container_pk = partition_key.PartitionKey(path=pk_definition.get('path'),
+                                                      kind=pk_definition.get('kind'),
+                                                      version=pk_definition.get('version'))
+            hashed_key = container_pk.get_effective_partition_key_string([partition_key_value])
```

Comment:
Just a note here, in .NET we ran into some trouble using the effective partition key string over an effective partition key range with queries for subpartitioning with prefix partition key. Since this is bulk and not query it might be a non-issue here.